test/onnx/test_utility_funs.py

```diff
@@ -802,6 +802,46 @@ def forward(self, x):
         funcs = onnx_model.functions
         self.assertIn("M", [f.name for f in funcs])
 
+    @skipIfUnsupportedMinOpsetVersion(15)
+    def test_local_function_predefined_attributes(self):
+        class M(torch.nn.Module):
+            num_layers: int
```

Comment:
These are very good questions. I found the explanation/definition for class annotations here https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations. 
> Does this need to be defined here rather than just in __init__? If so, why?

Yes, because [get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) tries to `Return a dictionary containing type hints for a function, method, module or class object`, not module/class instance.
And in pep-526, setting annotation inside `__init__` alone sets annotation for instance attribute, instead of class attribute.

> Does it have to do with type hints?

Yes, this is explained in pep-526.

> Does setting a type hint in __init__ not work?

No, same as first question. 

> If type hints are required, why are they?
> I would imagine the tracer would assign a type to the attribute even if there's no type hint?

Tracer does not pick up these attributes. We pick up these attributes by inspecting type of module in forward pre hook. So it's purely python and unrelated to tracer nor PyTorch. 

> Could you document this in the doc string for export_modules_as_functions in torch.onnx.export?

Adding.